core/src/main/java/io/grpc/internal/MessageFramer.java

```diff
@@ -342,7 +340,7 @@ private BufferChainOutputStream() {
 
     @Override
     public void write(int b) throws IOException {
-      singleByte[0] = (byte) b;
+      byte[] singleByte = new byte[]{(byte)b};
```

Comment:
> The JVM is heavily optimized for short lived objects

Hotspot is. Android is not. This code is used on Android.

If the BufferChainOutputStream instance was set to `null` on framer.close() then existing byte[] would have a similar lifetime, since most RPCs are unary.

> There is hope that the variable could be stack allocated, not a guarantee. It just means a better best-case scenario.

Yes, things may be better than worst-case, which is good if we are okay with the worst-case. But if we aren't okay with the worst-case and the hope is of unknown likelihood, then I don't see the value of the hope.

> This change is a trade off. We are optimizing global memory usage, at the cost of single byte writes.

The cost of single byte writes is severe enough that it should possibly be prohibited, because if you are calling it you won't be happy. Allocating for every byte is similar to `throw new UnsupportedOperationException()` in my mind because the user's reaction to it should be similar. It's an unacceptable implementation, at least at face-value.